File: src/Solutions/Euler129.hs
module Solutions.Euler129 where
import Utils.Operators((<:>))
-- A(n) < n

a k | gcd k 10 == 1 = let a' acc n | acc `mod` k == 0 = n
                                   | otherwise = a' ((10 * acc + 1) `mod` k) (n + 1)
                      in a' 1 1
    | otherwise = -1

euler129 :: IO String
euler129 = return $ show $ snd $ head $ dropWhile ((<=10^6) . fst) $ map (a <:> id) [10^6 + 1..]

File: rust/src/solutions/euler129.rs
fn a(k: i64) -> i64 {
    if num::integer::gcd(k, 10) == 1 {
        fn a_g(k: i64, acc: i64, n: i64) -> i64 {
            if acc % k == 0 {
                n
            } else {
                a_g(k, (10 * acc + 1) % k, n + 1)
            }
        }
        a_g(k, 1, 1)
    } else {
        -1
    }
}

pub fn euler129() -> String {
    // A(n) < n
    (1000001..).find(|&n| a(n) > 1000000).unwrap().to_string()
}
